src/coreclr/jit/lowerxarch.cpp

```diff
@@ -6314,38 +6314,92 @@ void Lowering::ContainCheckHWIntrinsic(GenTreeHWIntrinsic* node)
+                                    // 132 form: XMM1 = (XMM1 * [XMM3]) + XMM2
+                                    MakeSrcContained(node, op2);
+                                }
+                                else if (IsContainableHWIntrinsicOp(node, op3, &supportsRegOptional))
+                                {
+                                    // op1 = (op1 * op2) + [op3]
+                                    // 213 form: XMM1 = (XMM2 * XMM1) + [XMM3]
+                                    MakeSrcContained(node, op3);
 
-                            if (!HWIntrinsicInfo::CopiesUpperBits(intrinsicId))
+                                }
+                                else
+                                {
+                                    assert(supportsRegOptional);
+                                    // op1 = (op1 * op2) + [op3]
+                                    // 213 form: XMM1 = (XMM2 * XMM1) + [XMM3]
+                                    op3->SetRegOptional();
+                                }
+                                break;
+                            }
+                            case 3:
                             {
-                                // 231 form: op3 = (op2 * op3) + [op1]
-                                MakeSrcContained(node, op1);
+                                // 231 form: XMM1 = (XMM2 * [XMM3]) + XMM1
+                                // One of the following:
+                                // op3 = ([op1] * op2) + op3
+                                // op3 = (op1 * [op2]) + op3
+                                if (IsContainableHWIntrinsicOp(node, op1, &supportsRegOptional) && !HWIntrinsicInfo::CopiesUpperBits(intrinsicId))
+                                {
+                                    // Intrinsics with CopyUpperBits semantics cannot have op1 be contained
+
+                                    MakeSrcContained(node, op1);
+                                }
+                                else if (IsContainableHWIntrinsicOp(node, op2, &supportsRegOptional))
+                                {
+                                    MakeSrcContained(node, op2);
+                                }
+                                else
+                                {
+                                    assert(supportsRegOptional);
+                                    op2->SetRegOptional();
+                                }
+                                break;
                             }
-                        }
-                        else
-                        {
-                            assert(supportsRegOptional);
+                            default:
+                            {
+                                assert(overwrittenOpNum == 2 || overwrittenOpNum == 0);
+                                if (IsContainableHWIntrinsicOp(node, op1, &supportsRegOptional) && !HWIntrinsicInfo::CopiesUpperBits(intrinsicId))
```

Comment:
This likewise isn't handling the case where `overwritteOpNum == 0` and `IsContainableHWIntrinsicOp(node, op2, &supportsRegOptional)`.

I also wonder if there is something we can do here to better handle `overwrittenOpNum == 0` in general. In that scenario, any operand can be `regOptional` except LSRA doesn't support marking more than one as such, iirc.

We should still, however, be able to preference a form based on both `which operands are contained` and `which operands are last use`

Reply:
Do we have preference on the form when ```overwrittenOpNum == 0```? In this case, ```which operands are contained``` would be the only condition we need to check, right?

Reply:
I think that's covered by this comment: https://github.com/dotnet/runtime/pull/58196#discussion_r706316530

We can still make an informed choice and provide better codegen by biasing towards which operand is "last use", even if the destination is not one of the overwritten operands.